frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/IncrementalMojoHelper.java

```diff
@@ -0,0 +1,371 @@
+                // looking around inside can take a loooong time, we don't
+                // have to sacrifice much "correctness" and can lose the
+                // occasional rebuild when things are modified back to how
+                // they were. This assumes that mtime bubbles.
+                String lastModifiedTime = Long.toString(file.toAbsolutePath().toFile().lastModified());
+                files.add(new Execution.File(file.toString(), 0, lastModifiedTime));
+            }
+
+            return FileVisitResult.CONTINUE;
+        }
+
+        /**
+         * PERF NOTES:
+         * <ul>
+         *     <li>Yes, we're mixing the walking with the work,  but due to the
+         *     underlying library calls, we're 300ms better off in JSM DC</li>
+         *     <li>Yes, this removes parallelism, but even in JSM DC (worst case),
+         *     it's negligible, while for Stash and Jira DC it's faster to be single
+         *     threaded. Combined with these all usually being run in Maven
+         *     reactors, this is more likely to pay off by allowing for better
+         *     overall parallelism</li>
+         * </ul>
+         */
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+            String fileName = file.getFileName().toString();
+
+            if (excludedFilenames.contains(fileName)) {
+                return FileVisitResult.CONTINUE;
+            }
+
+            if (DIGEST_FILES.contains(fileName) ||
+                    DIGEST_EXTENSIONS.contains(getFileExtension(fileName))) {
+                addTrackedFile(files, file);
+            }
+
+            return FileVisitResult.CONTINUE;
+        }
+
+        @Override
+        public FileVisitResult visitFileFailed(Path path, IOException e) throws IOException {
+            log.debug("Failed to visit {}", path, e);
+            return super.visitFileFailed(path, e);
+        }
+
+        private static String getFileExtension(String fileName) {
+            int dotIndex = fileName.lastIndexOf('.');
+            if (dotIndex > 0 // skip over dot-files like .babelrc
+                    // check if the '.' is the last character == no extension
+                    && dotIndex < fileName.length() - 1) {
+                return fileName.substring(dotIndex + 1);
+            } else {
+                return null;
+            }
+        }
+    }
+
+    private Set<Execution.File> createFilesDigest() throws IOException {
+        final Set<Execution.File> files = new HashSet<>();
+
+        IncrementalVisitor visitor = new IncrementalVisitor(files, excludedFilenames);
+        Files.walkFileTree(workingDirectory.toPath(), visitor);
+        triggerFiles.forEach(file -> addTrackedFile(files, file.toPath()));
+
+        return files;
+    }
+
+    private static void addTrackedFile(Collection<Execution.File> files, Path file) {
+        try {
+            byte[] fileBytes = Files.readAllBytes(file);
```

Comment:
What if the file is 120Gbs big? ;)

Reply:
This method can't handle more than 2GB :P since primitive arrays can only have 2**31 values. In all seriousness, if we do have large files it'll get freed pretty quickly, this is in young gen.

Reply:
Yes - the point is that it will not get freed but rather fail if we happen to stumble upon something strange - but we can worry about that when it happens ;)